#### Abrir Dataset para su exploración


#### Importamos las librerías necesarias para la exploración


In [ ]:
import pandas as pd

In [2]:
db = pd.read_csv(r"C:\Users\mchap\Downloads\olist_limpio\fact_orders.csv")

In [20]:
# Exploración de los datos

print("Primeras 5 filas del dataset")
display(db.head(5))
print("----------------------------------------------------------------------")
print("Forma del dataset en columnas y filas")
display(db.shape)
print("----------------------------------------------------------------------")
print("sample de 5 filas del dataset")
display(db.sample(5, random_state=42))
print("----------------------------------------------------------------------")
print("Información del dataset")
db.info()
print("----------------------------------------------------------------------")
print("Descripción estadística del dataset")
display(db.describe(include='all'))
print("----------------------------------------------------------------------")
print("Tipo de datos de cada columna")
display(db.dtypes)  
print("----------------------------------------------------------------------")
print("Numero de valores nulos por columna")
display(db.isnull().sum())
print("----------------------------------------------------------------------")
print("Duplicados en el dataset")
display(db.duplicated().sum())


Primeras 5 filas del dataset


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,n_reviews,dias_aprobacion,dias_a_transportista,dias_entrega_total,dias_entrega_estimada,dias_retraso,entrega_tardia,fecha_compra,tiene_items,tiene_review
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,1.0,0.007431,2.373924,8.436574,15.544063,-7.107488,0.0,2017-10-02,True,True
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,1.0,1.279745,1.742627,13.782037,19.137766,-5.355729,0.0,2018-07-24,True,True
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,1.0,0.011505,0.216100,9.394213,26.639711,-17.245498,0.0,2018-08-08,True,True
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,7c142cf63193a1473d2e66489a9ae977,59296,...,1.0,0.012419,3.758252,13.208750,26.188819,-12.980069,0.0,2017-11-18,True,True
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,72632f0f9dd73dfee390c9b22eb56dd6,9195,...,1.0,0.042940,0.936053,2.873877,12.112049,-9.238171,0.0,2018-02-13,True,True


----------------------------------------------------------------------
Forma del dataset en columnas y filas


(99441, 33)

----------------------------------------------------------------------
sample de 5 filas del dataset


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,n_reviews,dias_aprobacion,dias_a_transportista,dias_entrega_total,dias_entrega_estimada,dias_retraso,entrega_tardia,fecha_compra,tiene_items,tiene_review
52263,b9a6c5f5df52c7226ac85aee7524c27f,f160aaf480efdfa7268f0fa535f73e76,delivered,2018-06-12 20:07:44,2018-06-12 20:44:26,2018-06-13 13:09:00,2018-06-19 12:44:08,2018-07-17,2d1bf256227e4d22d10ea6c0b81809d7,96900,...,1.0,0.025486,0.709213,6.691944,34.161296,-27.469352,0.0,2018-06-12,True,True
46645,261e71d2349c713eafa9f3df5972b95d,d6708bbbd2d419475869a84e41f620a1,delivered,2018-01-20 12:15:57,2018-01-20 12:37:13,2018-01-25 21:42:52,2018-01-30 11:32:35,2018-02-15,12bf514b8d413d8cbe66a2665f4b724c,30840,...,1.0,0.014769,5.393692,9.969884,25.488924,-15.519039,0.0,2018-01-20,True,True
37546,67b50899f52995848c427e361e10dde3,1b353c00c71689afba44554e43cc5a76,delivered,2018-06-16 21:24:10,2018-06-16 21:36:59,2018-06-21 13:55:00,2018-06-27 13:17:27,2018-07-16,83c6df0d47130de38c99cebe96521e8a,80220,...,1.0,0.008900,4.688079,10.662002,29.108218,-18.446215,0.0,2018-06-16,True,True
94756,32733fc014b67ef70fa6039dd8c6ba82,ad66699beab381d0747605f108cc02b4,delivered,2017-08-30 21:12:28,2017-08-31 02:50:24,2017-09-12 20:16:46,2017-09-25 17:53:23,2017-09-22,29b186723b197669f69b7d63c3e27c07,21044,...,1.0,0.234676,12.961319,25.861748,22.116343,3.745405,1.0,2017-08-30,True,True
14771,39a70e9e9b729b11dee34ac12478597f,20c678bfce3c8252288f70dd2bee5e51,delivered,2017-08-10 21:26:25,2017-08-10 21:44:13,2017-08-11 19:11:16,2017-08-22 16:45:00,2017-09-12,a59129ed35da4c3e3f2a005b4c6582fc,86081,...,1.0,0.012361,0.906146,11.804572,32.106655,-20.302083,0.0,2017-08-10,True,True


----------------------------------------------------------------------
Información del dataset
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 33 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   order_id                       99441 non-null  str    
 1   customer_id                    99441 non-null  str    
 2   order_status                   99441 non-null  str    
 3   order_purchase_timestamp       99441 non-null  str    
 4   order_approved_at              99281 non-null  str    
 5   order_delivered_carrier_date   97658 non-null  str    
 6   order_delivered_customer_date  96476 non-null  str    
 7   order_estimated_delivery_date  99441 non-null  str    
 8   customer_unique_id             99441 non-null  str    
 9   customer_zip_code_prefix       99441 non-null  int64  
 10  customer_city                  99441 non-null  str    
 11  customer_state        

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,n_reviews,dias_aprobacion,dias_a_transportista,dias_entrega_total,dias_entrega_estimada,dias_retraso,entrega_tardia,fecha_compra,tiene_items,tiene_review
count,99441,99441,99441,99441,99281,97658,96476,99441,99441,99441.000000,...,98673.000000,99281.000000,97658.000000,96476.000000,99441.000000,96476.000000,96476.000000,99441,99441,99441
unique,99441,99441,8,98875,90733,81018,95664,459,96096,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,634,2,2
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-03-31 15:08:21,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20,8d50f5eadf50201ccdcedfb9e2ac8455,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2017-11-24,True,True
freq,1,1,96478,3,9,47,3,522,17,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1176,98666,98673
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,35137.474583,...,1.005584,0.434129,3.234050,12.558702,23.767650,-11.179120,0.081129,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,29797.938996,...,0.075060,1.084917,3.611996,9.546530,8.832371,10.186113,0.273035,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1003.000000,...,1.000000,0.000000,-171.212419,0.533414,1.648993,-146.016123,0.000000,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,11347.000000,...,1.000000,0.008958,1.128967,6.766403,18.331690,-16.244384,0.000000,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,24416.000000,...,1.000000,0.014306,2.204676,10.217755,23.240370,-11.948941,0.000000,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,58900.000000,...,1.000000,0.607535,4.071777,15.720327,28.424861,-6.390000,0.000000,NaN,NaN,NaN


----------------------------------------------------------------------
Tipo de datos de cada columna


order_id                             str
customer_id                          str
order_status                         str
order_purchase_timestamp             str
order_approved_at                    str
order_delivered_carrier_date         str
order_delivered_customer_date        str
order_estimated_delivery_date        str
customer_unique_id                   str
customer_zip_code_prefix           int64
customer_city                        str
customer_state                       str
n_items                          float64
n_vendedores                     float64
total_price                      float64
total_freight                    float64
pago_total                       float64
n_metodos_pago                   float64
cuotas_max                       float64
tipo_pago_principal                  str
review_score                     float64
review_comment_message               str
review_creation_date                 str
n_reviews                        float64
dias_aprobacion 

----------------------------------------------------------------------
Numero de valores nulos por columna


order_id                             0
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                  160
order_delivered_carrier_date      1783
order_delivered_customer_date     2965
order_estimated_delivery_date        0
customer_unique_id                   0
customer_zip_code_prefix             0
customer_city                        0
customer_state                       0
n_items                            775
n_vendedores                       775
total_price                        775
total_freight                      775
pago_total                           1
n_metodos_pago                       1
cuotas_max                           1
tipo_pago_principal                  1
review_score                       768
review_comment_message           58605
review_creation_date               768
n_reviews                          768
dias_aprobacion                    160
dias_a_transportista     

----------------------------------------------------------------------
Duplicados en el dataset


np.int64(0)

#### Verificaremos order_id y granularidad


In [23]:
print("Cantidad de filas:", db.shape[0])
print("Pedidos únicos:", db["order_id"].nunique())
print("¿Hay una fila por pedido?:", db.shape[0] == db["order_id"].nunique())
print("order_id repetidos:", db["order_id"].duplicated().sum())

Cantidad de filas: 99441
Pedidos únicos: 99441
¿Hay una fila por pedido?: True
order_id repetidos: 0


In [26]:
# Debemos convertir las 5 columnas temporales a tipo datetime para poder trabajar con ellas y calcular la variable objetivo.

lista = [
    "order_purchase_timestamp",
    "order_approved_at", 
    "order_delivered_carrier_date", 
    "order_delivered_customer_date", 
    "order_estimated_delivery_date"
    
    ]

for col in lista:
    db[col] = pd.to_datetime(db[col], errors='coerce')


In [27]:
db[lista].dtypes
db[lista].isnull().sum()

order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [28]:
db["order_status"].value_counts(dropna=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [29]:
db_modelo = db.loc[
    (db["order_status"] == "delivered")
    & (db["order_delivered_customer_date"].notna())
    & (db["order_estimated_delivery_date"].notna())
].copy()

print(db_modelo.shape)

(96470, 33)


#### Crearemos el TARGET oficial del Proyecto llamado "pedido_retrasado"

In [30]:
db_modelo["pedido_retrasado"] = (
    db_modelo["order_delivered_customer_date"].dt.normalize()
    > db_modelo["order_estimated_delivery_date"].dt.normalize()
).astype(int)

In [32]:
db_modelo["pedido_retrasado"].value_counts()
print("Cantidad de pedidos entregados a tiempo:", db_modelo["pedido_retrasado"].value_counts()[0])
print("Cantidad de pedidos entregados con retraso:", db_modelo["pedido_retrasado"].value_counts()[1])

Cantidad de pedidos entregados a tiempo: 89936
Cantidad de pedidos entregados con retraso: 6534


In [33]:
# Calcular la tasa de pedidos retrasados
tasa_retraso = db_modelo["pedido_retrasado"].mean()

print(f"Tasa de pedidos retrasados: {tasa_retraso:.2%}")

Tasa de pedidos retrasados: 6.77%


In [34]:
db_modelo["retraso_con_hora"] = (
    db_modelo["order_delivered_customer_date"]
    > db_modelo["order_estimated_delivery_date"]
).astype(int)

tasa_con_hora = db_modelo["retraso_con_hora"].mean()

print(f"Tasa comparando fecha y hora: {tasa_con_hora:.2%}")

Tasa comparando fecha y hora: 8.11%


In [35]:
pd.crosstab(
    db_modelo["pedido_retrasado"],
    db_modelo["retraso_con_hora"],
    rownames=["Regla por día"],
    colnames=["Regla por fecha y hora"]
)

Regla por fecha y hora,0,1
Regla por día,,
0,88644,1292
1,0,6534


6.534 retrasos reales por día calendario → 6,77%
+ 1.292 falsos retrasos por comparar la hora
= 7.826 retrasos → 8,11%

Esto quiere decir que: 
8,11%: considera retrasado todo pedido cuya entrega ocurrió después de la fecha y hora prometidas.
6,77%: considera retrasado solo si la entrega ocurrió en un día calendario posterior al prometido.

El target oficial será pedido_retrasado, definido comparando días calendario.

In [ ]:
# Eliminamos la columna "retraso_con_hora" ya que no la vamos a utilizar en el modelo y solo nos interesa la variable objetivo "pedido_retrasado".
db_modelo = db_modelo.drop(columns="retraso_con_hora")

#### El siguiente paso será el control Anti-leakeage: revisaremos las columnas y separaremos cuales pueden ser variables predictoras y cuales estarán prohibidas

In [38]:
db_modelo.columns.tolist()

['order_id',
 'customer_id',
 'order_status',
 'order_purchase_timestamp',
 'order_approved_at',
 'order_delivered_carrier_date',
 'order_delivered_customer_date',
 'order_estimated_delivery_date',
 'customer_unique_id',
 'customer_zip_code_prefix',
 'customer_city',
 'customer_state',
 'n_items',
 'n_vendedores',
 'total_price',
 'total_freight',
 'pago_total',
 'n_metodos_pago',
 'cuotas_max',
 'tipo_pago_principal',
 'review_score',
 'review_comment_message',
 'review_creation_date',
 'n_reviews',
 'dias_aprobacion',
 'dias_a_transportista',
 'dias_entrega_total',
 'dias_entrega_estimada',
 'dias_retraso',
 'entrega_tardia',
 'fecha_compra',
 'tiene_items',
 'tiene_review',
 'pedido_retrasado']

Crearemos la primera feature valida: dias_prometidos

Mide cuántos días calendario tenía Olist para entregar desde la compra. Es una variable importante porque una promesa demasiado corta puede aumentar el riesgo de demora. Tendremos en consideración únicamente el día y desestimaremos la hora para no distorsionar los cálculos.

In [39]:
db_modelo["dias_prometidos"] = (
    db_modelo["order_estimated_delivery_date"].dt.normalize()
    - db_modelo["order_purchase_timestamp"].dt.normalize()
).dt.days

In [40]:
db_modelo["dias_prometidos"].describe()

count    96470.000000
mean        24.372738
std          8.758428
min          3.000000
25%         19.000000
50%         24.000000
75%         29.000000
max        156.000000
Name: dias_prometidos, dtype: float64

In [ ]:
# Queremos comprobar que no haya valores negativos o extraños antes de continuar con las demás variables.
print("Pedidos con 0 o menos días prometidos:", (db_modelo["dias_prometidos"] <= 0).sum())

Pedidos con 0 o menos días prometidos: 0


In [42]:
db_modelo["dias_prometidos"].describe()

count    96470.000000
mean        24.372738
std          8.758428
min          3.000000
25%         19.000000
50%         24.000000
75%         29.000000
max        156.000000
Name: dias_prometidos, dtype: float64

La variable es consistente y nos deja una primera lectura clara:

Medida	 Resultado	    Interpretación
Mínimo	 3 días	        No hubo promesas inmediatas o imposibles.
Promedio 24,37 días	    En promedio, Olist prometía entregar en unos 24 días.
Mediana	 24 días	    El pedido “típico” tenía 24 días de promesa.
25%–75%	 19 a 29 días	La mitad central de los pedidos tenía promesas relativamente concentradas.
Máximo	 156 días	    Es un valor muy alto que debemos investigar.

Los 156 días es un valor muy alto que investigaremos.

In [43]:
db_modelo.nlargest(10, "dias_prometidos")[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_estimated_delivery_date",
        "dias_prometidos",
        "total_price",
        "total_freight",
        "customer_state"
    ]
]

,order_id,order_purchase_timestamp,order_estimated_delivery_date,dias_prometidos,total_price,total_freight,customer_state
57160,eec7f369423b033e549c02f3c5381205,2018-02-06 20:44:56,2018-07-12,156,252.99,19.02,RS
40094,0607f0efea4b566f1eb8f7d3c2397320,2018-03-06 09:47:07,2018-08-03,150,125.98,20.18,SP
15791,c72727d29cde4cf870d569bf65edabfd,2017-02-07 18:01:15,2017-07-04,147,25.90,26.09,MA
86444,c2bb89b5c1dd978d507284be78a04cb2,2017-05-23 22:28:36,2017-10-11,141,199.98,122.88,SP
67488,40dc2ba6f322a17626aac6244332828c,2017-10-05 21:39:05,2018-01-30,117,40.00,14.10,RJ
14578,0d5ce32fbaa548555dac4bee840d7705,2017-07-01 15:47:06,2017-10-19,110,189.80,76.42,RN
37224,5aaba6681a3a926180ccd6ffaeaf6bd8,2017-12-25 23:45:35,2018-04-06,102,179.90,45.34,SC
92724,865aee0fd5873d8a2974df861e2526ef,2018-01-07 20:49:50,2018-04-17,100,179.90,45.34,SC
8210,09c636e254bb18eeef943452017249f4,2017-10-27 17:03:40,2018-02-02,98,169.90,70.41,MT
59011,65a7c3111cd734466eb706d544e6fae4,2017-03-23 17:36:38,2017-06-28,97,194.70,33.18,SP


Crearemos dos variables temporales a partir de la fecha de compra.

In [44]:
db_modelo["mes_compra"] = db_modelo["order_purchase_timestamp"].dt.month

db_modelo["dia_semana_compra"] = (
    db_modelo["order_purchase_timestamp"].dt.dayofweek
)

Interpretaremos dia_semana_compra como: 

0 = lunes
1 = martes
2 = miércoles
3 = jueves
4 = viernes
5 = sábado
6 = domingo

In [47]:
print("Mes de compra")
db_modelo["mes_compra"].value_counts().sort_index()

Mes de compra


mes_compra
1      7819
2      8208
3      9549
4      9101
5     10294
6      9231
7     10028
8     10544
9      4151
10     4743
11     7288
12     5514
Name: count, dtype: int64

In [48]:
print("Día de la semana de compra (0=lunes, 6=domingo)")
db_modelo["dia_semana_compra"].value_counts().sort_index()

Día de la semana de compra (0=lunes, 6=domingo)


dia_semana_compra
0    15701
1    15502
2    15074
3    14322
4    13684
5    10555
6    11632
Name: count, dtype: int64

Comprobaremos si dias_prometidos realmente esta relacionado con la demora.

In [49]:
db_modelo.groupby("pedido_retrasado")["dias_prometidos"].agg(
    ["count", "mean", "median", "min", "max"]
)

,count,mean,median,min,max
pedido_retrasado,,,,,
0,89936,24.451377,24.0,3,156
1,6534,23.290328,23.0,3,70


Esto compara los días prometidos entre:
- Pedidos puntuales (0).
- Pedidos retrasados (1).